wip: otel config envvar support (#3048)
grafana/alloy: docs/sources/set-up/migrate/from-otelcol.md

  Make sure that you use the new metric names, for example, in your alerts and dashboards queries.
* The logs produced by {{< param "PRODUCT_NAME" >}} differ from those produced by OpenTelemetry Collector.
* {{< param "FULL_PRODUCT_NAME" >}} exposes the {{< param "PRODUCT_NAME" >}} [UI][].

> * Environment variables with a scheme other than `env` aren't supported. Environment variables with no scheme default to `env`.

Reviewer: They can :( that being said the docs strongly suggest not to do any other scheme and I think they would need to use the collector builder to change the scheme.

> * Environment variables with a scheme other than `env` aren't supported. Environment variables with no scheme default to `env`.

Reviewer 2: but aren't other schemes other types of variables such as file path? like: ${file:exporters.yaml}
I believe that with the scheme "file", the collector does not resolve it as an environment variable but embeds the config that's found at the provided filepath?

> * Environment variables with a scheme other than `env` aren't supported. Environment variables with no scheme default to `env`.

Author: @wildum I don't fully understand the ins and outs of this flow yet, but I can outline what I've observed.

- The ConfigProvider takes a set of configmap provider factories. Each of these factories defines a `schemeName` accessible via `Scheme()`
- When the config file is parsed, if a value is encountered that contains a scheme (e.g. `${env:THING}`, `${file:THING}`, `${yaml:THING}`), the parser will route this value to the cfgmap provider registered for that `Scheme()` and give it a chance to process the value. <- This is what `val` is in the arguments to `Retrieve`.
- In the case of `env`, that is now routed to my custom provider, which essentially escapes the value (e.g. `${env:THING}` -> `$${env:THING}`) so that the rest of the parser ignores it. (Then later on I can do a regex replace with the appropriate Alloy syntax.)
- In the case of a different scheme, say `${file:WOW}`, my custom env provider is never called. What ends up happening is that the otel config parser spits out an error like `(Critical) failed to get otelcol config: cannot resolve the configuration: scheme "file" is not supported for uri "file:THING"`. It's not a great error message, but I can "fix" it by implementing yet another cfgmap provider which handles the `file:` scheme.
- If I switch over to the current state of main branch and modify one of the test files to include `file:`, I get the same error. So from what I can tell these changes don't improve or worsen the way that scheme (or others) are handled.

> * Environment variables with a scheme other than `env` aren't supported. Environment variables with no scheme default to `env`.

Author: This is where my knowledge kind of runs out, lol. From what I know, I agree with what you said; the conditional on the scheme name will never actually get used. My starting point, however, was [this](https://github.com/open-telemetry/opentelemetry-collector/blob/main/confmap/provider/envprovider/provider.go), which does include that logic. Possibly because it gets reused elsewhere?

For the case of no scheme being specified, that's where [this line](https://github.com/grafana/alloy/blob/edbc39c581e6bf09dbb8a3fd46a5cd1bd390e124/internal/converter/internal/otelcolconvert/otelcolconvert.go#L116) comes into play. Basically just adding `env:` to any non-qualified variable name. That was based on a recommendation from the OTel code here: https://github.com/open-telemetry/opentelemetry-collector/blob/eb00c827127868e3ee92dbff1b6259a32b486657/confmap/resolver.go#L56

> * Environment variables with a scheme other than `env` aren't supported. Environment variables with no scheme default to `env`.

Reviewer 2: ok thanks, then it's fine to keep it. Maybe you could add some of the knowledge that you shared here as a comment in the file?

> * Environment variables with a scheme other than `env` aren't supported. Environment variables with no scheme default to `env`.

Author: Done!

> * Environment variables with a scheme other than `env` are not supported. Environment variables with no scheme will default to `env`.

Reviewer 3: ```suggestion
* Environment variables with a scheme other than `env` aren't supported. Environment variables with no scheme default to `env`.
```

We try to use contractions on phrases like are not, cannot, it is, and so on.

We try to use present tense and active sentences, so words like "will" should be avoided. It's not a hard rule.. just a guideline/suggestion.
